File: Utilities/Matlab/scripts/mesh_transformation.m
% McDermott and McGrattan
% 10-4-12
% mesh_transformation.m
%
% Postprocessing script to extract transformation mapping from SMV.

close all
clear all

% Read SMV file and store transformations

datadir = '../../Verification/Miscellaneous/';
fid  = fopen([datadir,'mesh_transformation.smv'],'r');

I = 50;
J = 50;
LX = 1.5;
LY = 1.5;
dx = LX/I;
dy = LY/J;

while 1
    tline = fgetl(fid);
    if ~ischar(tline), break, end
    if strcmp(tline,'TRNX')
        %disp(tline)
        for i=1:3
            skip=fgetl(fid);
        end
        for i=1:I+1
            tline=fgetl(fid);
            C=textscan(tline,'%f');
            %disp(C{1})
            cx(i) = C{1}(1)*dx;
            x(i) = C{1}(2);
        end
    end
    if strcmp(tline,'TRNY')
        %disp(tline)
        for j=1:3
            skip=fgetl(fid);
        end
        for j=1:J+1
            tline=fgetl(fid);
            C=textscan(tline,'%f');
            %disp(C{1})
            cy(j) = C{1}(1)*dy;
            y(j) = C{1}(2);
        end
    end
end
fclose(fid);

% Make the plots

% From the mesh_transformation.fds input file:
CC = [0.3 1.2];
PC = [0.5 1.0];

plotdir = '../../Manuals/FDS_User_Guide/SCRIPT_FIGURES/';
plot_style
Title_Font_Size=20;

XMinorTick = 0.1:0.1:1.4;

plot(cx,x,'k-','LineWidth',2); hold on
for i=1:length(XMinorTick)
    % piece-wise linear function
    if XMinorTick(i)<CC(1)
        YMinorTick = (PC(1)/CC(1))*XMinorTick(i);
    elseif XMinorTick(i)<CC(2)
        YMinorTick = PC(1) + (PC(2)-PC(1))/(CC(2)-CC(1))*(XMinorTick(i)-CC(1));
    elseif XMinorTick(i)<=LX
        YMinorTick = PC(2) + (LX-PC(2))/(LX-CC(2))*(XMinorTick(i)-CC(2));
    end
    plot([XMinorTick(i) XMinorTick(i)],[0 YMinorTick],'k-')
    plot([0 XMinorTick(i)],[YMinorTick YMinorTick],'k-')
end
set(gca,'XTick',[0:0.3:1.5])
set(gca,'YTick',[0:0.3:1.5])
xlabel('\xi','FontSize',Title_Font_Size,'FontName',Font_Name,'Interpreter',Font_Interpreter)
ylabel('\itx','FontSize',Title_Font_Size,'FontName',Font_Name,'Interpreter',Font_Interpreter)

set(gca,'FontSize',Label_Font_Size)
set(gca,'FontName',Font_Name)
set(gca,'Units',Plot_Units)
set(gca,'Position',[Scat_Plot_X Scat_Plot_Y Scat_Plot_Width Scat_Plot_Height])

% print to pdf

set(gcf,'Visible',Figure_Visibility);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Scat_Paper_Width Scat_Paper_Height]);
set(gcf,'PaperPosition',[0 0 Scat_Paper_Width Scat_Paper_Height]);
print(gcf,'-dpdf',[plotdir,'piece_wise_linear_trnx'])

figure

plot(cy,y,'k-','LineWidth',2); hold on
for i=1:length(XMinorTick)
    % polynomial function
    YMinorTick = 2*XMinorTick(i) - 2*XMinorTick(i)^2 + 0.8889*XMinorTick(i)^3;
    plot([XMinorTick(i) XMinorTick(i)],[0 YMinorTick],'k-')
    plot([0 XMinorTick(i)],[YMinorTick YMinorTick],'k-')
end
set(gca,'XTick',[0:0.3:1.5])
set(gca,'YTick',[0:0.3:1.5])
xlabel('\xi','FontSize',Title_Font_Size,'FontName',Font_Name,'Interpreter',Font_Interpreter)
ylabel('\itx','FontSize',Title_Font_Size,'FontName',Font_Name,'Interpreter',Font_Interpreter)

set(gca,'FontSize',Label_Font_Size)
set(gca,'FontName',Font_Name)
set(gca,'Units',Plot_Units)
set(gca,'Position',[Scat_Plot_X Scat_Plot_Y Scat_Plot_Width Scat_Plot_Height])

% print to pdf

set(gcf,'Visible',Figure_Visibility);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Scat_Paper_Width Scat_Paper_Height]);
set(gcf,'PaperPosition',[0 0 Scat_Paper_Width Scat_Paper_Height]);
print(gcf,'-dpdf',[plotdir,'polynomial_trnx'])

File: Utilities/Python/scripts/mesh_transformation.py

# Postprocessing script to extract transformation mapping from SMV and create figures for User's Guide

import numpy as np
import matplotlib.pyplot as plt
import fdsplotlib

plot_style = fdsplotlib.get_plot_style('fds')

outdir = '../../Verification/Miscellaneous/'
pltdir = '../../Manuals/FDS_User_Guide/SCRIPT_FIGURES/'

fid = open(outdir + 'mesh_transformation.smv', 'r')

I = 50
J = 50
LX = 1.5
LY = 1.5
dx = LX / I
dy = LY / J

cx = np.zeros(I + 1)
x = np.zeros(I + 1)
cy = np.zeros(J + 1)
y = np.zeros(J + 1)

while True:
    tline = fid.readline()
    if not tline:  # End of file
        break
    tline = tline.strip()
    if tline == 'TRNX':
        # Skip next 3 lines
        for i in range(3):
            skip = fid.readline()
        # Read I+1 lines of transformation data
        for i in range(I + 1):
            tline = fid.readline()
            values = [float(val) for val in tline.split()]
            cx[i] = values[0] * dx
            x[i] = values[1]
    if tline == 'TRNY':
        # Skip next 3 lines
        for j in range(3):
            skip = fid.readline()
        # Read J+1 lines of transformation data
        for j in range(J + 1):
            tline = fid.readline()
            values = [float(val) for val in tline.split()]
            cy[j] = values[0] * dy
            y[j] = values[1]

fid.close()

# From the mesh_transformation.fds input file:
CC = [0.3, 1.2]
PC = [0.5, 1.0]

XMinorTick = np.arange(0.1, 1.5, 0.1)

# First figure - piece-wise linear transformation

fig = fdsplotlib.plot_to_fig(x_data=cx, y_data=x, marker_style='k-',
                             x_min=0, x_max=1.5, y_min=0, y_max=1.5,
                             figure_size=(plot_style['Scat_Paper_Width'],plot_style['Scat_Paper_Height']),
                             plot_size=(plot_style['Scat_Plot_Width'],plot_style['Scat_Plot_Height']),
                             plot_origin=(plot_style['Scat_Plot_X'],plot_style['Scat_Plot_Y']),
                             x_label=r'$\xi$', y_label=r'$x$')

for i in range(len(XMinorTick)):
    if XMinorTick[i] < CC[0]:
        YMinorTick = (PC[0] / CC[0]) * XMinorTick[i]
    elif XMinorTick[i] < CC[1]:
        YMinorTick = PC[0] + (PC[1] - PC[0]) / (CC[1] - CC[0]) * (XMinorTick[i] - CC[0])
    elif XMinorTick[i] <= LX:
        YMinorTick = PC[1] + (LX - PC[1]) / (LX - CC[1]) * (XMinorTick[i] - CC[1])

    fdsplotlib.plot_to_fig(x_data=[XMinorTick[i],XMinorTick[i]], y_data=[0,YMinorTick], marker_style='k-', figure_handle=fig)
    fdsplotlib.plot_to_fig(x_data=[0, XMinorTick[i]], y_data=[YMinorTick,YMinorTick],   marker_style='k-', figure_handle=fig)

ax = plt.gca()
ax.set_xticks(np.arange(0, 1.5, 0.3))
ax.set_yticks(np.arange(0, 1.5, 0.3))

plt.savefig(pltdir + 'piece_wise_linear_trnx.pdf', format='pdf')
plt.close()

# Second figure - polynomial transformation

fig = fdsplotlib.plot_to_fig(x_data=cy, y_data=y, marker_style='k-',
                             x_min=0, x_max=1.5, y_min=0, y_max=1.5,
                             figure_size=(plot_style['Scat_Paper_Width'],plot_style['Scat_Paper_Height']),
                             plot_size=(plot_style['Scat_Plot_Width'],plot_style['Scat_Plot_Height']),
                             plot_origin=(plot_style['Scat_Plot_X'],plot_style['Scat_Plot_Y']),
                             x_label=r'$\xi$', y_label=r'$x$')

for i in range(len(XMinorTick)):
    YMinorTick = 2 * XMinorTick[i] - 2 * XMinorTick[i]**2 + 0.8889 * XMinorTick[i]**3
    fdsplotlib.plot_to_fig(x_data=[XMinorTick[i],XMinorTick[i]], y_data=[0,YMinorTick], marker_style='k-', figure_handle=fig)
    fdsplotlib.plot_to_fig(x_data=[0, XMinorTick[i]], y_data=[YMinorTick,YMinorTick],   marker_style='k-', figure_handle=fig)

ax = plt.gca()
ax.set_xticks(np.arange(0, 1.5, 0.3))
ax.set_yticks(np.arange(0, 1.5, 0.3))

plt.savefig(pltdir + 'polynomial_trnx.pdf', format='pdf')
plt.close()

